java/ql/lib/semmle/code/java/controlflow/Guards.qll

```diff
@@ -175,14 +174,45 @@ class Guard extends ExprParent {
   }
 }
 
+/**
+ * A `Guard` that tests an expression's type -- that is, an `instanceof T` or a
+ * `case T varname` pattern case.
+ */
+class TypeTestGuard extends Guard {
+  Expr testedExpr;
+  Type testedType;
+
+  TypeTestGuard() {
+    exists(InstanceOfExpr ioe | this = ioe |
+      testedExpr = ioe.getExpr() and
+      testedType = ioe.getCheckedType()
+    )
+    or
+    exists(PatternCase pc | this = pc |
+      pc.getSelectorExpr() = testedExpr and
+      testedType = pc.getPattern().getType()
+    )
+  }
+
+  /** Holds if this guard tests whether `e` has type `t`. */
+  predicate appliesTypeTest(Expr e, Type t) { e = testedExpr and t = testedType }
+}
+
 private predicate switchCaseControls(SwitchCase sc, BasicBlock bb) {
```

Comment:
I don't think we need to account for `PatternCase` here - rather I think it should be excluded. It ought to be handled by the regular `ConditionBlock.controls` disjunct in the `directlyControls` predicate above.

Reply:
Agreed, this is redundant now pattern cases are recognised as condition blocks. Removed. I have however adapted this to note that the predecessor might not be the top of the switch block any longer -- it could be a previous pattern case (and added tests for this).